js/ui/filter_builder/filter_builder.js

```diff
@@ -0,0 +1,798 @@
+    FILTER_BUILDER_ITEM_VALUE_TEXT_CLASS = "dx-filterbuilder-item-value-text",
+    FILTER_BUILDER_OVERLAY_CLASS = "dx-filterbuilder-overlay",
+    ACTIVE_CLASS = "dx-state-active";
+
+var ACTIONS = [
+        "onEditorPreparing", "onEditorPrepared"
+    ],
+    OPERATORS = {
+        and: "And",
+        or: "Or",
+        notAnd: "!And",
+        notOr: "!Or"
+    };
+
+var EditorFactory = Class.inherit(EditorFactoryMixin);
+
+var FilterBuilder = Widget.inherit({
+    _getDefaultOptions: function() {
+        return extend(this.callBase(), {
+            /**
+              * @name dxFilterBuilderOptions_onEditorPreparing
+              * @publicName onEditorPreparing
+              * @type function(e)
+              * @type_function_param1 e:object
+              * @type_function_param1_field1 value:any
+              * @type_function_param1_field2 setValue(newValue):any
+              * @type_function_param1_field3 cancel:boolean
+              * @type_function_param1_field4 editorElement:jQuery
+              * @type_function_param1_field6 editorName:string
+              * @type_function_param1_field7 editorOptions:object
+              * @type_function_param1_field8 dataField:string
+              * @extends Action
+              * @action
+             */
+            onEditorPreparing: null,
+
+            /**
+              * @name dxFilterBuilderOptions_onEditorPrepared
+              * @publicName onEditorPrepared
+              * @type function(e)
+              * @type_function_param1 e:object
+              * @type_function_param1_field1 value:any
+              * @type_function_param1_field2 setValue(newValue):any
+              * @type_function_param1_field3 cancel:boolean
+              * @type_function_param1_field4 editorElement:jQuery
+              * @type_function_param1_field6 editorName:string
+              * @type_function_param1_field7 editorOptions:object
+              * @type_function_param1_field8 dataField:string
+              * @extends Action
+              * @action
+             */
+            onEditorPrepared: null,
+
+            /**
+            * @name dxFilterBuilderOptions_fields
+            * @publicName fields
+            * @type Array<Object>
+            * @default []
+            */
+            fields: [],
+
+            /**
+            * @name dxFilterBuilderOptions_fields_caption
+            * @publicName caption
+            * @type string
+            * @default undefined
+            */
+
+            /**
+            * @name dxFilterBuilderOptions_fields_dataField
+            * @publicName dataField
+            * @type string
+            * @default undefined
+            */
+
+            /**
+             * @name dxFilterBuilderOptions_fields_dataType
+             * @publicName dataType
+             * @type string
+             * @default "string"
+             * @acceptValues "string" | "number" | "date" | "boolean" | "object"
+             */
```

Comment:
What for must use the "object" type?

Reply:
dxFilterBuilder has fields with nested items. "object" type supports only 2 operations: isblank and isnotblank